### ISSUE TITLE
CMake configuration error if all ENABLE_ options are set to OFF

### ISSUE THREAD

**Issue by PeterBowman** [roboticslab-uc3m/kinematics-dynamics#104]
Regression caused by #14.

Steps to reproduce this issue:

1. `git clone` the repository
2. create a `build` folder and configure with all options set to OFF (as in [asibot-hmi/.travis.yml](https://github.com/roboticslab-uc3m/asibot-hmi/blob/d6d9e49/.travis.yml#L27)):
   ```
   cmake .. -DENABLE_TrajectoryLib:BOOL=OFF \
            -DENABLE_KdlVectorConverterLib:BOOL=OFF \
            -DENABLE_BasicCartesianControl:BOOL=OFF \
            -DENABLE_CartesianControlServer:BOOL=OFF \
            -DENABLE_CartesianControlClient:BOOL=OFF \
            -DENABLE_KdlSolver:BOOL=OFF \
            -DENABLE_BasicTwoLimbCartesianControl:BOOL=OFF \
            -DENABLE_TwoLimbCartesianControlServer:BOOL=OFF \
            -DENABLE_transCoordsUsingJoints:BOOL=OFF \
            -DENABLE_tests:BOOL=OFF
   ```

Symptom:

```
bartek@bartek-ubuntu:~/Desktop/kinematics-dynamics/build$ cmake ..
-- YCM found in /usr/local/share/YCM.
-- Setting up installation of roboticslab-kinematics-dynamics.ini to /usr/local/share/yarp/config/path.d folder.
--  --- plugin BasicCartesianControl: disabled
--  --- plugin BasicTwoLimbCartesianControl: disabled
--  --- plugin CartesianControlClient: disabled
--  --- plugin CartesianControlServer: disabled
--  --- plugin KdlSolver: disabled
--  --- plugin TwoLimbCartesianControlServer: disabled
-- Configuring done
CMake Error: INSTALL(EXPORT) given unknown export "ROBOTICSLAB_KINEMATICS_DYNAMICS"
-- Generating done
-- Build files have been written to: /home/bartek/Desktop/kinematics-dynamics/build
```

Additionally: https://travis-ci.org/roboticslab-uc3m/asibot-hmi/builds/256413744.

Cause: CMake tries to export a set of targets named `ROBOTICSLAB_KINEMATICS_DYNAMICS`. However, it's unable to find a `install(TARGETS ... EXPORT ROBOTICSLAB_KINEMATICS_DYNAMICS)` call since everything is disabled.

Remark: there are two sets of targets exported in parallel - one for the build-tree by means of the `export(INSTALL)` command, and another one for the install-tree by `install(EXPORT)`. Both are currently bundled within YCM's `InstallBasicPackageFiles` module ([docs](http://robotology.github.io/ycm/gh-pages/master/module/InstallBasicPackageFiles.html), [source](https://github.com/robotology/ycm/blob/b271d6d8fb3ec52f0881c97935439331e8c59659/modules/InstallBasicPackageFiles.cmake), [local command invocation](https://github.com/roboticslab-uc3m/kinematics-dynamics/blob/8c1e7b04e1ea9471794634626fc28c96ed55a2e6/CMakeLists.txt#L128-L133)). The former does no harm, it just generates tons of boilerplate at `<pkg>Targets.cmake`. Sadly, the latter is not so permisible and throws an error.

**Current workaround:** enable at least one of the following exported library targets:

* `-DENABLE_TrajectoryLib:BOOL=ON`
* `-DENABLE_KdlVectorConverterLib:BOOL=ON`
* `-DENABLE_Gait:BOOL=ON` (you need to build [Gait](https://github.com/roboticslab-uc3m/gait) on your system OR fetch kin-dyn code with `git clone --recursive`)

**Comment by PeterBowman** [comment]
A new target, `ROBOTICSLAB::YarpTinyMathLib`, was added in the meantime to the export set, so it might work now (at least until the CLI CMake options are set to ignore this target, too). Anyway, I tried to export our dummy `ROBOTICSLAB_KINEMATICS_DYNAMICS` library, solely intended to generate a export header. These lines do the trick when importing targets from the **build tree**:

```cmake
install(TARGETS ROBOTICSLAB_KINEMATICS_DYNAMICS
        EXPORT ROBOTICSLAB_KINEMATICS_DYNAMICS
        DESTINATION ${CMAKE_BINARY_DIR}
        OPTIONAL)
```

Thus, we don't even need to install an actual library file since it's marked as `OPTIONAL` and `add_library` uses `EXCLUDE_FROM_ALL`. That is, issuing a plain `make` command (no build target) prevents from compiling our dummy `empty.c`, linking it into a shared library file and, considering the code above, copying such file to a system location upon `make install`.

There is one minor caveat and one deal-breaker for this solution:

* We try to install a library target that may not even physically exist on disk as a `.dll`/`.so`. Starting from CMake 3.6, the `EXCLUDE_FROM_ALL` parameter will silence an annoying warning related to this fact.
* This won't work with targets pulled from the **install tree** due to a generic check intended to ensure that every imported file associated to a target actually exists on disk, throwing an error otherwise. This is an excerpt from `build/CMakeFiles/Export/lib/cmake/ROBOTICSLAB_KINEMATICS_DYNAMICS/ROBOTICSLAB_KINEMATICS_DYNAMICSTargets-release.cmake`:
  ```cmake
  # Import target "ROBOTICSLAB::ROBOTICSLAB_KINEMATICS_DYNAMICS" for configuration "Release"
  set_property(TARGET ROBOTICSLAB::ROBOTICSLAB_KINEMATICS_DYNAMICS APPEND PROPERTY IMPORTED_CONFIGURATIONS RELEASE)
  set_target_properties(ROBOTICSLAB::ROBOTICSLAB_KINEMATICS_DYNAMICS PROPERTIES
    IMPORTED_LOCATION_RELEASE "/home/bartek/git/roboticslab/kinematics-dynamics/build/libROBOTICSLAB_KINEMATICS_DYNAMICS.so"
    IMPORTED_SONAME_RELEASE "libROBOTICSLAB_KINEMATICS_DYNAMICS.so"
  )
  
  list(APPEND _IMPORT_CHECK_TARGETS ROBOTICSLAB::ROBOTICSLAB_KINEMATICS_DYNAMICS )
  list(APPEND _IMPORT_CHECK_FILES_FOR_ROBOTICSLAB::ROBOTICSLAB_KINEMATICS_DYNAMICS "/home/bartek/git/roboticslab/kinematics-dynamics/build/libROBOTICSLAB_KINEMATICS_DYNAMICS.so" )
  ```
  The `_IMPORT_CHECK_FILES_FOR_<target>` will break things on a `find_package` call for downstreams.

**Comment by PeterBowman** [comment]
Solved at 0ff614d. I gave up on using the [`InstallBasicPackageFiles`](http://robotology.github.io/ycm/gh-pages/master/module/InstallBasicPackageFiles.html) module from YCM. It did provide a massive shortcut to standard CMake commands, although on the other hand we needed to pass the project version as a required parameter (and our `<pkg>Cmake.config.in` had to be moved to the root path).

BTW minimum required CMake version was bumped to 3.1.3. For Trusty users, please make sure you have installed the [`cmake3` package](https://packages.ubuntu.com/trusty/cmake3).

**Comment by PeterBowman** [comment]
> BTW minimum required CMake version was bumped to 3.1.3. For Trusty users, please make sure you have installed the cmake3 package.

I switched back to 2.8.11 (280c300), didn't realize that YCM provides a set of enhanced CMake modules (future releases pulled from GH and custom patches). Tested and working.
